### PR TITLE
[Proposal] demo compressor

### PR DESCRIPTION
**Pull request by LeonardoWang**
now we provide compressors for pytorch and Tensorflow, including 4 compress algorithms:

### Sensitivity Pruner
**Learning both Weights and Connections for Efficient Neural Networks**
https://arxiv.org/pdf/1506.02626.pdf

### Auto Gradual Pruner
**To prune, or not to prune: exploring the efficacy of pruning for model compression**
https://arxiv.org/pdf/1710.01878.pdf

### Quantize Aware Training Quantizer
**Quantization and Training of Neural Networks for Efficient Integer-Arithmetic-Only Inference**
https://arxiv.org/pdf/1712.05877.pdf

### DoReFa Quantizer
**DOREFA-NET: TRAINING LOW BITWIDTH CONVOLUTIONAL NEURAL NETWORKS WITH LOW BITWIDTH GRADIENTS**
https://arxiv.org/pdf/1606.06160.pdf

For pytorch, we can just add new algorithm in it's forward method to instrument a layer
For tensorflow, we can modify graph by swapping output of new weight and old weight
To implement pruner, we use mask and update weight information
To implement quantizer, we use just replace old weight data with new one

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -10,6 +10,11 @@ jobs:
   steps:
   - script: python3 -m pip install --upgrade pip setuptools --user
     displayName: 'Install python tools'
+  - script: |
+      python3 -m pip install torch==0.4.1 --user
```
Comment (leckie-chn) [inline azure-pipelines.yml]: suggest using `python3 -m pip install -r requirement.txt`

File: azure-pipelines.yml [hunk 2]
```diff
@@ -50,6 +55,11 @@ jobs:
   steps:
   - script: python3 -m pip install --upgrade pip setuptools
     displayName: 'Install python tools'
+  - script: |
+      python3 -m pip install torch==0.4.1 --user
```
Comment (leckie-chn) [inline azure-pipelines.yml]: same as above

File: docs/en_US/Compressor/Overview.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+NNI provides easy-to-use toolkit to help user  design and use compression algorithm.
+
+## Framework
+We use the instrumentation method to insert a node or function after the corresponding position in the model.
+<br>
+When compression algorithm designer implements one prune algorithm, he only need to pay attention to the generation method of mask, without caring about applying the mask to the garph.
+## algorithm
+We now provide some naive compression algorithm and four popular compress agorithms for users, including two pruning algorithm and two quantization algorithm.
+Below is a list of model compression algorithms supported in our compressor
+|    Name           |  Paper    |
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Overview.md]: this table cannot be correctly rendered

File: docs/en_US/Compressor/Overview.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
+# Compressor
+NNI provides easy-to-use toolkit to help user  design and use compression algorithm.
+
+## Framework
+We use the instrumentation method to insert a node or function after the corresponding position in the model.
+<br>
+When compression algorithm designer implements one prune algorithm, he only need to pay attention to the generation method of mask, without caring about applying the mask to the garph.
+## algorithm
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Overview.md]: Algorithm

File: docs/en_US/Compressor/Overview.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+Pytorch code
+```
+nni.compressors.torchCompressor.LevelPruner(sparsity=0.8).compress(model)
+```
+
+Our compressor will automatically insert mask into your model, and you can train your model with masks without changing your training code. You will get a compressed model when you finish your training.
+
+You can get more information in Algorithm details
+
+
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Overview.md]: add another section to show how to customize a new pruning or quantize algorithm

File: docs/en_US/Compressor/Overview.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# Compressor
+NNI provides easy-to-use toolkit to help user  design and use compression algorithm.
+
+## Framework
+We use the instrumentation method to insert a node or function after the corresponding position in the model.
+<br>
+When compression algorithm designer implements one prune algorithm, he only need to pay attention to the generation method of mask, without caring about applying the mask to the garph.
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Overview.md]: typo: garph -> graph

Comment (chicm-ms) [inline docs/en_US/Compressor/Overview.md]: 'he' :  the designer maybe a female.

File: setup.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def read(fname):
 
 setup(
     name = 'nni',
-    version = '999.0.0-developing',
+    version = 'v0.8-263-g1f92408',
```
Comment (chicm-ms) [inline setup.py]: It seems this change is made by makefile, do not checkin this change.

File: src/sdk/pynni/nni/compressors/example/main_tf_pruner.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
+from nni.compressors.tfCompressor import AGPruner
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compressors/example/main_tf_pruner.py]: suggest change 'main' in file names to 'mnist'

File: docs/en_US/Compressor/Overview.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+## Usage
+
+Take naive level pruner as an example
+
+If you want to prune all weight to 80% sparsity, you can add code below into your code before your training code.
+
+Tensorflow code
+```
+nni.compressors.tfCompressor.LevelPruner(sparsity=0.8).compress(model_graph)
```
Comment (chicm-ms) [inline docs/en_US/Compressor/Overview.md]: format (camel case) of package names 'tfCompressor' and 'torchCompressor' are inconsistent with current nni package names.

File: src/sdk/pynni/nni/compressors/tfCompressor/quantizer.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            a = tf.math.tanh(weight)
+            b = a/(2*tf.reduce_max(tf.abs(weight))) + 0.5
+
+            scale = pow(2, self.q_bits-1)
+            with tf.get_default_graph().gradient_override_map({'Round': 'Identity'}):
+                qw = tf.round(b*scale)/scale
+            r_qw = 2*qw - 1
+            return r_qw
+
+except ImportError:
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compressors/tfCompressor/quantizer.py]: Why catch ImportError?

File: src/sdk/pynni/nni/compressors/example/main_tf_pruner.py (hunk 8 above)
Comment (LeonardoWang) [reply src/sdk/pynni/nni/compressors/example/main_tf_pruner.py]: it maybe confusing with the class Mnist

File: setup.py [hunk 11]
```diff
@@ -58,7 +58,8 @@ def read(fname):
         'schema',
         'PythonWebHDFS',
         'colorama',
-        'sklearn'
+        'sklearn',
+        'pyyaml'
```
Comment (liuzhe-lz) [inline setup.py]: We have already required `ruamel.yaml`.
Should not add another YAML library.

File: src/sdk/pynni/nni/compressors/example/main_tf_pruner.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
+from nni.compressors.tf_compressor import AGPruner
```
Comment (liuzhe-lz) [inline src/sdk/pynni/nni/compressors/example/main_tf_pruner.py]: Examples should be placed in `examples` folder, not `sdk`.

File: src/sdk/pynni/requirements.txt [hunk 13]
```diff
@@ -8,3 +8,6 @@ hyperopt
 
 # metis tuner
 sklearn
+
+# compresser
+pyyaml
```
Comment (liuzhe-lz) [inline src/sdk/pynni/requirements.txt]: Use `ruamel.yaml`

File: src/sdk/pynni/setup.py [hunk 14]
```diff
@@ -36,7 +36,8 @@ def read(fname):
         'json_tricks',
         'numpy',
         'scipy',
-        'coverage'
+        'coverage',
+        'pyyaml'
```
Comment (liuzhe-lz) [inline src/sdk/pynni/setup.py]: Same

File: src/webui/yarn.lock [hunk 15]
```diff
@@ -3,73 +3,72 @@
```
Comment (liuzhe-lz) [inline src/webui/yarn.lock]: Checkout this file from master

File: setup.py [hunk 16]
```diff
@@ -27,7 +27,7 @@ def read(fname):
 
 setup(
     name = 'nni',
-    version = '999.0.0-developing',
+    version = 'v0.8-347-g0538d30',
```
Comment (QuanluZhang) [inline setup.py]: change version back

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    def preprocess_model(self, model):
+        """
+        This method is called when a model is bound to the compressor.
+        Users can optionally overload this method to do model-specific initialization.
+        It is guaranteed that only one model will be bound to each compressor instance.
+        """
+        pass
+
+
+class TfLayerInfo:
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: add docstring for this class

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    file = open(config_path, 'r')
+    yaml = YAML(typ='safe')
+    yaml_text = yaml.load(file.read())
+    configure_file = yaml_text.get(class_name, {})
+    if not configure_file:
+        print('WARNING: load Nothing from configure file, Default { }')
+    return configure_file
+
+class TfPruner(TfCompressor):
+    """TODO"""
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: TODO what?

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+
+
+def _tf_detect_prunable_layers(model):
+    # search for Conv2D layers
+    # TODO: whitelist
+    # please make sure op not from optimizer
+    whiltlist = ['Conv2D', 'DepthwiseConv2dNative']
+    return [ TfLayerInfo(op) for op in model.get_operations() if op.type in whiltlist ]
+
+def _tf_default_get_configure(configure_list, layer_info):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: add docstring for functions

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    configure = {}
+    for config in configure_list:
+        if config.get('support_type', '') == 'default':
+            configure = config
+        elif layer_info.layer.type in config.get('support_type', []):
+            configure = config
+        elif layer_info.name in config.get('support_op', []):
+            configure = config
+
+    return configure
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: should properly check whether configure is `{}` when calling this function

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+
+def _tf_default_load_configure_file(config_path, class_name):
+    print('load CLASS:{0} from PATH:{1}'.format(class_name, config_path))
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: please using log rather than print

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        raise NotImplementedError("Pruners must overload calc_mask()")
+
+    def compress(self, model):
+        super().compress(model)
+        # TODO: configurable whitelist
+        for layer_info in _tf_detect_prunable_layers(model):
+            self._instrument_layer(layer_info)
+
+    def _instrument_layer(self, layer_info):
+        # it seems the graph editor can only swap edges of nodes or remove all edges from a node
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: using docstring here

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+        # we assume there is a proxy operation between the weight and the Conv2D layer
+        # this is true as long as the weight is `tf.Value`
+        # not sure what will happen if the weight is calculated from other operations
+        weight_op = layer_info.layer.inputs[layer_info.weight_index].op
+        weight = weight_op.inputs[0]
+        mask = self.calc_mask(layer_info, weight)
+        new_weight = weight * mask
+        tf.contrib.graph_editor.swap_outputs(weight_op, new_weight.op)
+    
+    def update_epoch(self, epoch, sess):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: add docstring

File: src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+        pass
+    
+    def step(self, sess):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/_nnimc_tf.py]: add docstring. And why `update_epoch` and `step` are in `TfPruner` rather than `TfCompressor`?

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+
+class LevelPruner(TfPruner):
+    def __init__(self, configure_list):
+        """
+            Configure Args:
+                sparsity
+        """
+        super().__init__()
+        self.configure_list = []
+        if isinstance(configure_list, list):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: do you accept config file path?

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+            for configure in configure_list:
+                self.configure_list.append(configure)
+        else:
+            raise ValueError('please init with configure list')
+
+    def load_configure(self, config_path):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: is this function called by others?

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+
+class AGPruner(TfPruner):
+    def __init__(self, configure_list):
+        """
+            Configure Args
+                initial_sparsity
+                final_sparsity
+                start_epoch
+                end_epoch
+                frequency
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: please give more description. what is the meaning of `frequency`?

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+    
+    def calc_mask(self, layer_info, weight):
+        sparsity = self.get_sparsity(_tf_default_get_configure(self.configure_list, layer_info))
+
+        threshold = tf.contrib.distributions.percentile(tf.abs(weight), sparsity * 100)
+        return tf.cast(tf.math.greater(tf.abs(weight), threshold), weight.dtype)
+
+class AGPruner(TfPruner):
+    def __init__(self, configure_list):
+        """
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: brief describe how this algorithm works

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+    
+    def compute_target_sparsity(self, layer_info):
+        configure = _tf_default_get_configure(self.configure_list, layer_info)
+        end_epoch = configure.get('end_epoch', 1)
+        start_epoch = configure.get('start_epoch', 0)
+        freq = configure.get('frequency', 1)
+        final_sparsity = configure.get('final_sparsity', 0)
+        initial_sparsity = configure.get('initial_sparsity', 0)
+
+        if end_epoch <= start_epoch or initial_sparsity >= final_sparsity:
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: is it an error if end_epoch < start_epoch?

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+        config_list = _tf_default_load_configure_file(config_path, 'AGPruner')
+        for config in config_list.get('config', []):
+            self.configure_list.append(config)
+
+    def calc_mask(self, layer_info, weight):
+        
+        target_sparsity = self.compute_target_sparsity(layer_info)
+        threshold = tf.contrib.distributions.percentile(weight, target_sparsity * 100)
+        mask = tf.stop_gradient(tf.cast(tf.math.greater(weight, threshold), weight.dtype))
+        print('tensor weight', weight)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: use log debug here if you think logging `weight` is helpful during debug

File: src/sdk/pynni/nni/compressors/tf_compressor/pruner.py [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+    def load_configure(self, config_path):
+        config_list = _tf_default_load_configure_file(config_path, 'AGPruner')
+        for config in config_list.get('config', []):
+            self.configure_list.append(config)
+
+    def calc_mask(self, layer_info, weight):
+        
+        target_sparsity = self.compute_target_sparsity(layer_info)
+        threshold = tf.contrib.distributions.percentile(weight, target_sparsity * 100)
+        mask = tf.stop_gradient(tf.cast(tf.math.greater(weight, threshold), weight.dtype))
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compressors/tf_compressor/pruner.py]: please explain why use `tf.stop_gradient` here in comments
